[SPARK-4969][STREAMING][PYTHON] Add binaryRecords to streaming

freeman-lab opened the pull request:
In Spark 1.2 we added a `binaryRecords` input method for loading flat binary data. This format is useful for numerical array data, e.g. in scientific computing applications. This PR adds support for the same format in Streaming applications, where it is similarly useful, especially for streaming time series or sensor data.

Summary of additions
- adding `binaryRecordsStream` to Spark Streaming 
- exposing `binaryRecordsStream` in the new PySpark Streaming
- new unit tests in Scala and Python

This required adding an optional Hadoop configuration param to `fileStream` and `FileInputStream`, but was otherwise straightforward.

@tdas @davies

freeman-lab committed:
Expose additional argument combination

freeman-lab committed:
Add binaryRecordsStream to python

freeman-lab committed:
Add binaryRecordStream to Java API

freeman-lab committed:
Add binaryRecordsStream to scala

freeman-lab committed:
Tests for binaryRecordsStream

freeman-lab committed:
Expose optional hadoop configuration

freeman-lab committed:
Fix missing arg

freeman-lab committed:
Reorganization

freeman-lab committed:
Add params to docstring

freeman-lab committed:
Spelling

freeman-lab committed:
Simpler default arg handling

freeman-lab committed:
Formatting

freeman-lab committed:
Line length

freeman-lab committed:
Add tests in python

freeman-lab committed:
Formatting

freeman-lab committed:
Formatting

davies commented:
The Python part look good to me, thanks.

freeman-lab commented on streaming/src/test/scala/org/apache/spark/streaming/InputStreamsSuite.scala:
```diff
@@ -233,6 +236,47 @@ class InputStreamsSuite extends TestSuiteBase with BeforeAndAfter {
+      Thread.sleep(1000)
+      // Set up the streaming context and input streams
+      val newConf = conf.clone.set(
+        "spark.streaming.clock", "org.apache.spark.streaming.util.SystemClock")
```
Ok great, I'll wait for your PR to be merged and then refactor this test accordingly.

JoshRosen replied:
My PR has been merged, so I think this should be unblocked now.

freeman-lab committed:
Make test inline

freeman-lab commented on streaming/src/test/scala/org/apache/spark/streaming/InputStreamsSuite.scala:
```diff
@@ -233,6 +236,47 @@ class InputStreamsSuite extends TestSuiteBase with BeforeAndAfter {
     }
   }
 
+  def testBinaryRecordsStream() {
```
Good catch, fixed.

SparkQA commented:
[Test build #25066 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25066/consoleFull) for   PR 3803 at commit [`317b6d1`](https://github.com/apache/spark/commit/317b6d1dc45f0706987c3258beaa64be08df4b3c).
 * This patch merges cleanly.

freeman-lab commented on streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala:
```diff
@@ -373,6 +393,25 @@ class StreamingContext private[streaming] (
+    val conf = sc_.hadoopConfiguration
+    conf.setInt(FixedLengthBinaryInputFormat.RECORD_LENGTH_PROPERTY, recordLength)
+    val br = fileStream[LongWritable, BytesWritable, FixedLengthBinaryInputFormat](directory, conf)
+    val data = br.map{ case (k, v) => v.getBytes}
```
Thanks for flagging this, I wasn't aware of that behavior of ``getBytes``. I think that, as you suggest, both here and in ``binaryRecords()`` it's not a problem in practice. The BytesWritable that comes from the FixedLengthBinaryInputFormat will always be backed by a Byte array that's of the fixed length. For consistency and good practice I'm happy to make the change from #2712 both here and in the other method. Or we just add a comment. Let me know which you'd prefer. One concern might be performance effects, as mentioned in the other PR? @sryza might have thoughts.

JoshRosen replied:
Probably no need to make a copy if this is actually safe (it would just harm performance), so I think it would be fine to add a comment.  I suppose we could also do an assert to check that the array isn't padded, since that's unlikely to be expensive.

freeman-lab commented:
Thanks for the review! I'll wait for @JoshRosen 's PR to merge and then update the test here. And will wait for your thoughts on the `getBytes` issue. Otherwise, I think everything's addressed.

SparkQA commented:
[Test build #25066 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25066/consoleFull) for   PR 3803 at commit [`317b6d1`](https://github.com/apache/spark/commit/317b6d1dc45f0706987c3258beaa64be08df4b3c).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class FileInputDStream[K, V, F <: NewInputFormat[K,V]](`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25066/
Test PASSed.

JoshRosen commented:
/bump, now that my test-refactoring PR has been merged.

freeman-lab commented:
Great thanks @JoshRosen will finish this up ASAP!

freeman-lab committed:
Merge remote-tracking branch 'upstream/master' into streaming-binary-records

freeman-lab committed:
Add note and defensive assertion for byte length

freeman-lab committed:
Refactor to reflect changes to FileInputSuite

freeman-lab commented:
@JoshRosen I finished the refactored tests and added better handling of the `getBytes` based on your suggestion.

SparkQA commented:
[Test build #26435 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26435/consoleFull) for   PR 3803 at commit [`9a3715a`](https://github.com/apache/spark/commit/9a3715a1e6a71040d234da52bf848b0bb109a591).
 * This patch merges cleanly.

JoshRosen commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -671,7 +675,10 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
       classOf[BytesWritable],
       conf=conf)
-    val data = br.map{ case (k, v) => v.getBytes}
+    val data = br.map{ case (k, v) =>
```
Super-minor style nit, but I think our house-style here is to have a space before the opening `{` and to place the closing `}` on its own line.

JoshRosen commented on streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala:
```diff
@@ -372,6 +392,33 @@ class StreamingContext private[streaming] (
+    val data = br.map{ case (k, v) =>
+      val bytes = v.getBytes
+      assert(bytes.length == recordLength, "Byte array does not have correct length")
+      bytes}
```
Same style nit applies in this file, too.

JoshRosen commented on streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaStreamingContext.scala:
```diff
@@ -210,6 +210,18 @@ class JavaStreamingContext(val ssc: StreamingContext) extends Closeable {
+   * @param directory HDFS directory to monitor for new files
+   * @param recordLength The length at which to split the records
+   */
+
```
This looks like an extra line of blank whitespace.

JoshRosen commented on streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala:
```diff
@@ -237,7 +240,14 @@ class FileInputDStream[K: ClassTag, V: ClassTag, F <: NewInputFormat[K,V] : Clas
     val fileRDDs = files.map(file =>{
-      val rdd = context.sparkContext.newAPIHadoopFile[K, V, F](file)
+      val rdd = conf match {
+        case Some(config) => context.sparkContext.newAPIHadoopFile(file,
```
Another minor style thing, but I'd put the `file` on its own line, like the other arguments.

JoshRosen commented:
@tdas pointed out that these new methods should be marked as `@Experimental`, since `binaryRecords` is experimental.

tdas commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -657,6 +657,10 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
    *
    * Load data from a flat binary file, assuming the length of each record is constant.
    *
+   * '''Note:''' Normally getBytes returns an array padded with extra values,
```
`getBytes` on what?

JoshRosen replied:
This is referring to the `BytesWritable` values: their `getBytes` methods are technically allowed to return padded byte arrays, so in general callers should take this into account, but in this case we are guaranteeing that `BytesWritable.getBytes` will return a non-padded array.

Actually, this raises a good point: the fact that we're returning un-padded arrays is an implementation detail, so maybe we don't want to make it part of our API contract (unless it's explicitly part of the FixedLengthBinaryInputFormat contract).

freeman-lab replied:
Hm, it's explicitly part of `FixedLengthBinaryInputFormat` in so far as that method constructs a bytes writable backed by a byte array of the given fixed record length ([here](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/input/FixedLengthBinaryRecordReader.scala#L111-123)). Does that count? In either case, I can expand the comment to avoid confusion.

tdas replied:
Does the end user ever need to call getBytes() given that this RDD is of type Array[Byte]? Does the user even need to know that this feature uses FixedLengthBinaryInputFormat underneath. And if they dont, then this note can be simplified and mentioned in terms of the RDD, something like "Each record in this DStream/RDD is of fixed length with padded bytes / right length / etc."

SparkQA commented:
[Test build #26435 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26435/consoleFull) for   PR 3803 at commit [`9a3715a`](https://github.com/apache/spark/commit/9a3715a1e6a71040d234da52bf848b0bb109a591).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class FileInputDStream[K, V, F <: NewInputFormat[K,V]](`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26435/
Test PASSed.

freeman-lab committed:
Space formatting

freeman-lab committed:
Formatting

freeman-lab committed:
Add experimental tag

SparkQA commented:
[Test build #26449 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26449/consoleFull) for   PR 3803 at commit [`14bca9a`](https://github.com/apache/spark/commit/14bca9aff82b7323a624c4e05a46b60dd140ec99).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26449 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26449/consoleFull) for   PR 3803 at commit [`14bca9a`](https://github.com/apache/spark/commit/14bca9aff82b7323a624c4e05a46b60dd140ec99).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class FileInputDStream[K, V, F <: NewInputFormat[K,V]](`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26449/
Test FAILed.

freeman-lab commented:
Jenkins, retest this please

SparkQA commented:
[Test build #26476 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26476/consoleFull) for   PR 3803 at commit [`14bca9a`](https://github.com/apache/spark/commit/14bca9aff82b7323a624c4e05a46b60dd140ec99).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26476 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26476/consoleFull) for   PR 3803 at commit [`14bca9a`](https://github.com/apache/spark/commit/14bca9aff82b7323a624c4e05a46b60dd140ec99).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class FileInputDStream[K, V, F <: NewInputFormat[K,V]](`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26476/
Test PASSed.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala:
```diff
@@ -361,6 +363,25 @@ class StreamingContext private[streaming] (
+   * @tparam V Value type for reading HDFS file
+   * @tparam F Input format for reading HDFS file
+   */
+  def fileStream[
```
Can you make these parameters a superset of the parameters in the other fileStream? Otherwise it seems like people can either use a conf or use a filter

freeman-lab replied:
Fine with me, though would require wiring the extra arguments (`filter` and `newFilesOnly`) through to all the binary record methods (because they'll also need the `conf`). Should we expose those arguments in `binaryRecords`? Or just use their defaults?

tdas replied:
Why the wiring to binaryRecords? Those two parameters are only relevant to
DStream, so nothing to pass on down to the RDDs
On Feb 2, 2015 7:30 PM, "Jeremy Freeman" <notifications@github.com> wrote:

> In
> streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala
> <https://github.com/apache/spark/pull/3803#discussion_r23980541>:
>
> > @@ -361,6 +363,25 @@ class StreamingContext private[streaming] (
> >
> >    /**
> >     * Create a input stream that monitors a Hadoop-compatible filesystem
> > +   * for new files and reads them using the given key-value types and input format.
> > +   * Files must be written to the monitored directory by "moving" them from another
> > +   * location within the same file system. File names starting with . are ignored.
> > +   * @param directory HDFS directory to monitor for new file
> > +   * @param conf Hadoop configuration
> > +   * @tparam K Key type for reading HDFS file
> > +   * @tparam V Value type for reading HDFS file
> > +   * @tparam F Input format for reading HDFS file
> > +   */
> > +  def fileStream[
>
> Fine with me, though would require wiring the extra arguments (filter and
> newFilesOnly) through to all the binary record methods (because they'll
> also need the conf). Should we expose those arguments in binaryRecords?
> Or just use their defaults?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/3803/files#r23980541>.
>

freeman-lab replied:
Sorry, meant exposing them to `binaryRecordsStream`. That's calling `fileStream`, and to include `conf` as an arg we'll also need to either specify `filter` and `newFilesOnly` or pass defaults, right?

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaStreamingContext.scala:
```diff
@@ -210,6 +211,18 @@ class JavaStreamingContext(val ssc: StreamingContext) extends Closeable {
   }
 
   /**
```
Please expose the new version of filestream in javaStreamingContext.

freeman-lab committed:
Add experimental tag

davies commented:
The Python parts look good to me, thanks!

freeman-lab committed:
Expose new version of fileStream with conf in java

freeman-lab committed:
Add filter and newFilesOnly alongside conf

- Ensures the argument including conf is a superset of existing
arguments
- Use default versions for binaryRecordStream

freeman-lab committed:
Add experimental tag

freeman-lab committed:
Simplify notes

SparkQA commented:
[Test build #26680 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26680/consoleFull) for   PR 3803 at commit [`eba925c`](https://github.com/apache/spark/commit/eba925c125341a1ed694c5f3ac3ab3914488f880).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26680 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26680/consoleFull) for   PR 3803 at commit [`eba925c`](https://github.com/apache/spark/commit/eba925c125341a1ed694c5f3ac3ab3914488f880).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class FileInputDStream[K, V, F <: NewInputFormat[K,V]](`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26680/
Test PASSed.

tdas commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -671,7 +674,11 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
+    val data = br.map { case (k, v) =>
+      val bytes = v.getBytes
+      assert(bytes.length == recordLength, "Byte array does not have correct length")
+      bytes
```
nit: Is this something that the user should be made aware of in the docs?

freeman-lab replied:
Do you mean something more than these notes we're adding? I just clarified the notes a bit to make it obvious the check is on the byte array.

tdas replied:
I meant should the user be told that the system can throw error when the records are not of the expected size. I dont have any strong feeling on this, just wondering.

freeman-lab replied:
Gotcha, I think it's ok as is then. Given what `FixedLengthInputFormat` is doing, this is more a defensive assertion, it's not something the user should hit due to an inappropriate input.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaStreamingContext.scala:
```diff
@@ -210,6 +211,20 @@ class JavaStreamingContext(val ssc: StreamingContext) extends Closeable {
+   * yielding byte arrays
+   * @param directory HDFS directory to monitor for new files
+   * @param recordLength The length at which to split the records
+   */
```
Shouldnt this have the note as well?

freeman-lab replied:
Thanks, added!

tdas commented:
LGTM. Just two minor nits.

freeman-lab committed:
Add note to java streaming context

freeman-lab committed:
Clarify note

freeman-lab commented:
Thanks for the detailed look @tdas! Think I addressed both nits.

SparkQA commented:
[Test build #26725 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26725/consoleFull) for   PR 3803 at commit [`b676534`](https://github.com/apache/spark/commit/b676534067a626260b6921ba17a04b6e03ff587a).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26725 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26725/consoleFull) for   PR 3803 at commit [`b676534`](https://github.com/apache/spark/commit/b676534067a626260b6921ba17a04b6e03ff587a).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class FileInputDStream[K, V, F <: NewInputFormat[K,V]](`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26725/
Test PASSed.

tdas commented:
Thanks Jeremy! Good clean patch!
On Feb 3, 2015 10:45 PM, "UCB AMPLab" <notifications@github.com> wrote:

> Test PASSed.
> Refer to this link for build results (access rights to CI server needed):
> https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26725/
> Test PASSed.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/3803#issuecomment-72799040>.
>
